### ISSUE TITLE
MakerNotes (Canon)

### ISSUE THREAD

**Issue by v1r0x** [pel/pel#98]
I switched from the "built-in" `exif_read_data` to PEL, but still missing the marker notes. I decided to write (at least for canon or at least for images from my camera) a parser for maker notes. I'm relatively new to the topic of how exif works and especially how makernotes work. Thus I wanted to share my findings and maybe someone can help with the stuff I don't understand (yet).
For the record: I have a Canon EOS 650D/Rebel T4i with Firmware 1.0.4 and Exif 2.3.
1. MakerNotes start with 2 Bytes of ...I don't know, maybe an identifier? In my case it is `0x0027`
2. Beside these 2 bytes they seem to have the same 12 byte structure as the rest of Exif. 2 Byte Tag ID, 2 Byte data type, 4 Byte length, 4 Byte value.
  - But: The value seems to always be an offset to the actual data. And the actual data seems to always be a 4 Byte long (at least for the values I checked which should be short instead of long), except for ascii (and probably float and double).
  - The offset is not a relative offset to the start of the IFD or something else, but an absolute offset.
My approach so far:
1. Before creating a new `PelEntryUndefined`in https://github.com/pel/pel/blob/master/src/PelIfd.php#L472 I parse the maker notes based on the `PelDataWindow` `$data`.
2. I start parsing after the first two bytes and looping over the `$data` in 12 byte blocks
```php
// New method in PelDataWindow
public function parseMakerNotes() {
    $idOffset = 2; // first two bytes, in my case 0x27
    // I know that $this->size is wrong, but suitable for testing
    for($i=0; $i<$this->size; $i++) {
        $offset = 12 * $i;
        $tag = $this->getShort($offset + $idOffset);
        $type = $this->getShort($offset + $idOffset + 2);
        $length = $this->getShort($offset + $idOffset + 4);
        $valOffset = $this->getShort($offset + $idOffset + 8);
        // since $valOffset is absolute we have to cut off the starting offset of the MakerNotes IFD
        $relOffset = $valOffset - $this->start;
    }
}
```
3. The problem is that `$relOffset` doesn't match the real offset where the data is stored. It is off by `0x1E`. If I add `0x1E` back to `$relOffset` it works like a charm.

Has anyone worked on this before or can provide any help?
I'm also not sure about where I should add my code and where and how I should parse the MakerNotes to match the current layout. If anyone is interested I could create a branch with my testing code and share it. Would love to help improving PEL :)

**Comment by canonknipser** [comment]
A good source for exif informations is allways exiftool: https://sno.phy.queensu.ca/~phil/exiftool/TagNames/Canon.html

**Comment by v1r0x** [comment]
thanks for the link. I already had a list of the tags. What I don't know and doesn't seem to be answered anywhere is the meaning of the first two bytes `0x0027`.
At least I figured out why I had to add `0x1E` to the offsets (see 3.). It's the length of JTIF header. The same as it is for other IFDs

**Comment by v1r0x** [comment]
Fixed by #99
